### PR TITLE
Refactor FileUploadUtils to use HttpClient 4.5.3

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
1. Make file upload from local and HDFS use the same HttpClient
2. Add optional headers and parameters into upload segment request
3. Remove the retry logic from FileUploadUtils, client should handle retry based on the response status code
4. Add exception class HttpErrorStatusException to pass both error message and status code
5. Move the file download logic into a new class FileDownloadUtils

### CODE REVIEW

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java [hunk 1]
```diff
@@ -1,470 +1,349 @@
...
-    public abstract EntityEnclosingMethod forUri(String uri);
+  public enum FileUploadType {
+    URI, JSON, TAR;
+
+    public static FileUploadType getDefaultUploadType() {
+      return TAR;
+    }
   }
 
-  public static int sendFile(final String host, final String port, final String path, final String fileName,
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: Need to check if third-eye is using some of these APIs

File: pinot-common/src/test/java/com/linkedin/pinot/common/utils/FileUploadUtilsTest.java [hunk 2]
```diff
@@ -15,79 +15,66 @@
  */
 package com.linkedin.pinot.common.utils;
 
+import com.linkedin.pinot.common.utils.FileUploadUtils.FileUploadType;
+import com.sun.net.httpserver.Headers;
```
Comment (mcvsubbu) [inline pinot-common/src/test/java/com/linkedin/pinot/common/utils/FileUploadUtilsTest.java]: com.sun?

File: pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java [hunk 3]
```diff
@@ -48,20 +48,24 @@ public void fetchSegmentToLocal(final String uri, final File tempFile) throws Ex
...
+          LOGGER.info("Downloaded file from {} to {}; Length of downloaded file: {}; Response status code: {}", uri,
+              tempFile, tempFile.length(), statusCode);
           return true;
-        } catch (PermanentDownloadException e) {
-          LOGGER.error("Failed to download file from {}, won't retry", uri, e);
-          throw e;
+        } catch (HttpErrorStatusException e) {
+          int statusCode = e.getStatusCode();
+          if (statusCode >= 500) {
+            LOGGER.error("Failed to download file from {}, might retry", uri, e);
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpSegmentFetcher.java]: These info should be logged from the retry policy, will address in a separate RB.

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileDownloadUtils.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  public static int downloadFile(String fileUri, File dest) throws Exception {
+    HttpUriRequest request = RequestBuilder.get(fileUri).setVersion(HttpVersion.HTTP_1_1).build();
+    try (CloseableHttpResponse response = HTTP_CLIENT.execute(request)) {
+      int statusCode = response.getStatusLine().getStatusCode();
+      if (statusCode >= 400) {
+        throw new HttpErrorStatusException(getErrorMessage(fileUri, response), statusCode);
+      }
+
+      HttpEntity entity = response.getEntity();
+      long contentLength = entity.getContentLength();
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileDownloadUtils.java]: Old one also checks the content length, so should be fine.

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileDownloadUtils.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    if (response.containsHeader(CommonConstants.Controller.HOST_HTTP_HEADER)) {
+      controllerHost = response.getFirstHeader(CommonConstants.Controller.HOST_HTTP_HEADER).getValue();
+      controllerVersion = response.getFirstHeader(CommonConstants.Controller.VERSION_HTTP_HEADER).getValue();
+    }
+    StatusLine statusLine = response.getStatusLine();
+    String errorMessage = String.format("Got error status code: %d with reason: %s while downloading file with uri: %s",
+        statusLine.getStatusCode(), statusLine.getReasonPhrase(), fileUri);
+    if (controllerHost != null) {
+      errorMessage =
+          String.format("%s from controller: %s, version: %s", errorMessage, controllerHost, controllerVersion);
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileDownloadUtils.java]: It's handled differently (reading from content vs directly get from the status line)

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java]: Good point, they do. Will update their code as well.

File: pinot-common/src/test/java/com/linkedin/pinot/common/utils/FileUploadUtilsTest.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-common/src/test/java/com/linkedin/pinot/common/utils/FileUploadUtilsTest.java]: They are used to mock a http server (not changed in this PR).

**Pull request merged**